extras/upgrade_fw.py

```diff
@@ -0,0 +1,161 @@
+        rc1, vendor_string, stderr1 = run_ssh(
+            dmi_chassis_vendor, raise_stderr=False
+        )
+        rc2, type_string, stderr2 = run_ssh(
+            dmi_chassis_type, raise_stderr=False
+        )
+
+        err_msg = ""
+        if rc1 != 0:
+            err_msg = vendor_string + stderr1 + "\n"
+        if rc2 != 0:
+            err_msg = err_msg + type_string + stderr2
+        if err_msg:
+            err_msg = (
+                "Unable to detect device vendor/type due to lacking of dmi info.\n"
+                + err_msg
+            )
+            logger.error(err_msg)
+            raise RuntimeError(err_msg)
+
+        type_index = int(type_string)
+        upgrade_type = Dmi.chassis_types[type_index]
+        try:
+            dev = [
+                dev
+                for dev in devices
+                if dev.fw_update_type in upgrade_type
+                and any(x == vendor_string for x in dev.vendor)
+            ][0]
+            logger.info("%s is a %s %s" % (ip, vendor_string, dev.__name__))
+        except IndexError:
+            err_msg = "Cannot find a proper Device class for %s: %s" % (
+                vendor_string,
+                Dmi.chassis_names[type_index],
+            )
+            logger.error(err_msg)
+            raise RuntimeError(err_msg)
+
+        if issubclass(dev, LVFSDevice):
+            return dev(ip, user, password)
+        elif issubclass(dev, OEMDevice):
+            if not (
+                "bmc_ip" in options
+                and "bmc_user" in options
+                and "bmc_password" in options
+            ):
+                raise RuntimeError(
+                    "Please provide $BMC_IP, $BMC_USER, $BMC_PASSWORD for this device"
+                )
+            exit()
+            return dev(
+                options.get("bmc_ip"),
+                options.get("bmc_user"),
+                options.get("bmc_password"),
+            )
+    except subprocess.CalledProcessError as e:
+        logger.error(e.output)
+        raise RuntimeError(e.output)
+
+
+def main():
+    """
+    mandantory parameter: $DEVICE_IP (testflinger env variable)
+    extra parameters for Server: $BMC_IP, $BMC_USERNAME, $BMC_PASSWORD
+    (not yet provided by testflinger)
+    """
+    parser = argparse.ArgumentParser(
+        epilog="Example: %(prog)s upgrade 10.0.0.1",
+        usage="%(prog)s {upgrade, downgrade, detect} $DEVICE_IP [options]",
+    )
+    parser.add_argument(
+        choices=["upgrade", "downgrade", "detect"],
+        dest="action",
+        help="select to upgrade firmware, downgrade firmware, or just detect DUT info without further actions",
+    )
+    parser.add_argument("device_ip", help="$DEVICE_IP")
+    parser.add_argument(
+        "--bmc_ip",
+        "-i",
+        nargs="?",
+    )
+    parser.add_argument(
+        "--bmc_user",
+        "-u",
+        nargs="?",
+    )
+    parser.add_argument(
+        "--bmc_password",
```

Comment:
I guess it’s easier in shell scripts to use hyphens instead of underscores.

Reply:
Done.